cmd/generic-handlers.go

```diff
@@ -469,3 +474,67 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 	h.handler.ServeHTTP(w, r)
 }
+
+// logginHandler logs all HTTP requests and responses
+// and write the log to a io.Writer object.
+type loggingHandler struct {
+	writer     io.Writer
+	handler    http.Handler
+	timeFormat string
+}
+
+func (h loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Log HTTP request
+	requestDump, err := httputil.DumpRequest(r, false)
+	if err == nil {
+		reqTime := time.Now().Format(h.timeFormat) + " >> "
+		h.writer.Write([]byte(reqTime))
+		h.writer.Write(requestDump)
+		h.writer.Write([]byte("\n"))
+	} else {
+		errorIf(err, "Cannot dump HTTP request")
+	}
+
+	// Setup a http recorder to save HTTP response
+	rec := httptest.NewRecorder()
+	h.handler.ServeHTTP(rec, r)
+
+	// Log HTTP Response
+	respDump, err := httputil.DumpResponse(rec.Result(), false)
+	if err == nil {
+		respTime := time.Now().Format(h.timeFormat) + " << "
+		h.writer.Write([]byte(respTime))
+		h.writer.Write(respDump)
+		h.writer.Write([]byte("\n\n"))
+	} else {
+		errorIf(err, "Cannot dump HTTP response")
+	}
+
+	// We copy the captured response headers to our new response
+	for k, v := range rec.Header() {
+		w.Header()[k] = v
+	}
+
+	// Grab the captured response body
+	w.WriteHeader(rec.Code)
+	w.Write(rec.Body.Bytes())
+}
+
+// Enable HTTP requests and responses logging handler
+func setLoggingHandler(h http.Handler) http.Handler {
+	logDir := os.Getenv("MINIO_TRACE_DIR")
+	if logDir == "" {
+		// Do nothing if no MINIO_TRACE_DIR is not set
+		return h
+	}
+	// Create the log file
+	logFile, err := os.Create(filepath.Join(logDir, "minio-"+time.Now().Format("2006-01-02T15:04:05")+".log"))
+	if err != nil {
+		errorIf(err, "Cannot create the log file")
+	}
+	return loggingHandler{
+		writer:     logFile,
+		handler:    h,
+		timeFormat: "2006-01-02 15:04:05 -0700",
```

Comment:
I think `timeFormat` field is not required.  You could use `"2006-01-02 15:04:05 -0700"` format inside the handler.